### PR TITLE
✨♻️ Big progress display revision

### PR DESCRIPTION
**Pull request by EvilGremlin**
### Description

Basically this is unification of print progress that makes it hardware-agnostic and all-granular, overall way more convenient to work with. All options are as independent as they can be, which allows for easy UI customisation.

[new_progress.webm](https://user-images.githubusercontent.com/22657714/190661590-684a2f5e-4191-4cad-bded-2d17d7870ede.webm)

### Benefits

M73 doesn't depend on display existence anymore.

`M73_REPORT` Prusa-like M73 that report values back. This allow hosts to get M73 progress from SD printing for somewhat accurate time estimates. Octoprint plugin: https://github.com/arekm/octopi_eta_override

`USE_M73_INTERACTION_TIME, SHOW_INTERACTION_TIME` Adds "time to user interaction". Prusa use it for filament change, but it can be anything, i.e. "pause at layer" - this will require some additional support from slicers.  

`PRINT_PROGRESS_SHOW_DECIMALS` now affect every UI progress display uniformly, and M73 too.

`ROTATE_PROGRESS_DISPLAY` is gone as it lost all meaning. However much progress string displays you enable can be rotated in any native UI, or no rotation if you leave only one. New function iterates over pointer array on `blink`, so each blink next string is redrawn. This also allow for completely arbitrary string placement on screen: all rotated in one place, all separate, in any groups... only thing to care about is to not leave artifacts from previous string/value. Same approach can be used for rotation of any screen element at all, i.e. multiple hotend temps on 1602 LCD, possibly even all in one array to save resources on problematic configurations, like AVR running delta kinematic.

### Bugfixes and small stuff

Added long overdue `fs.img` to gitignore.

`permyriadtostr4` now always 4-width.

Fixed progress bar drawing in LIGHTWEIGHT_UI ( thanks Scawty - "The code is relying on int being 16-bit so that the arithmetic right shift fills the left side with ones. On an ESP32 (and arm) it's 32-bit and not being sign extended" )

LIGHTWEIGHT_UI now redraw progress bar even less often which should be saving even more cycles.

### Things to do

Possibly need to expand capabilities report.

Some options to beep/flash lights when interaction time approach zero.

HD44780 strings need rewrite similar to LIGHTWEIGHT_UI to avoid artifacts when timer goes down from hundreds of hours.
`permyriadtostr4` need further touchups: whole single digit values don't show decimal, doesn't look nice.

### Related Issues

Draft with some notes: https://github.com/MarlinFirmware/Marlin/pull/24714

### CODE REVIEW

**Comment by arekm** [comment]
About capabilities report. Long time ago Marlin reported in M115 (a129078927c8968022569175cec7fea35c920fd2) which g-codes are supported in emergency parser like:

```
+    #define EMERGENCY_PARSER_CAPABILITIES " EMERGENCY_CODES:M108,M112,M410"
```

Later that was changed to support documented (on reprap.org wiki) EMERGENCY_PARSER capability in 3c9a838651861968109ea06b731d2a92905eb115 (which unfortunately doesn't tell host which commands are supported)
(https://reprap.org/wiki/Firmware_Capabilities_Protocol#Cap:EMERGENCY_PARSER)

Maybe it would be good to have a way to report which emergency codes are supported, like earlier EMERGENCY_CODES, and make that documented in https://reprap.org/wiki/Firmware_Capabilities_Protocol wiki for others to follow.

**Review by EvilGremlin — COMMENTED** [review]
I don't like prusa wording too, progmem not infinite... I think we should ask @arekm if he's ok with wording change, and @foosel in case she ever intend to support this natively

**Comment by arekm** [comment]
No need to follow prusa wording. There is only one plugin (from all plugins at http://plugins.octoprint.org/) that parses this and it will be updated to whatever wording is used, so pick best one from your perspective:

https://github.com/arekm/octopi_eta_override/tree/marlin2

Making two variants in one firmware is complicating things even more. Pick only one IMO.

(btw. OctoPrint way is kind of "if it can be done by plugin then there is no need for core to support it)

**Comment by The-EG** [comment]
Was this intended to be merged? I see the corresponding options have been changed in the configuration repo, but bugfix-2.1.x still expects at least some of the old options, ie. `Marlin/src/lcd/dogm/status_screen_DOGM.cpp` line 544:

```c++
#if BOTH(SHOW_REMAINING_TIME, ROTATE_PROGRESS_DISPLAY) 
```

That means building current bugfix-2.1.x with examples from the config repo (branch = bugfix-2.1.x) is now broken. It may build, but it's not going to show what it should on the LCD.


edit: this may be why the CI tests are failing too.

**Comment by thinkyhead** [comment]
> Was this intended to be merged?

Pretty soon. It still needs more code cleanup and review. Once merged, other PRs will be updated so tests can pass.

**Comment by EvilGremlin** [comment]
@thinkyhead with all the stuff you did in ABM, is it possible to use some pieces of it to grab all relevant stuff from example confs into tests?

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -1426,37 +1423,43 @@
...
 
-// LCD Print Progress options
-#if EITHER(SDSUPPORT, LCD_SET_PROGRESS_MANUALLY)
-  #if CAN_SHOW_REMAINING_TIME
-    //#define SHOW_REMAINING_TIME         // Display estimated time to completion
-    #if ENABLED(SHOW_REMAINING_TIME)
-      //#define USE_M73_REMAINING_TIME    // Use remaining time from M73 command instead of estimation
-      //#define ROTATE_PROGRESS_DISPLAY   // Display (P)rogress, (E)lapsed, and (R)emaining time
-    #endif
+// Add the G-code 'M73' to set / report the current job progress
```
Comment (thisiskeithb) [inline Marlin/Configuration_adv.h:1437]: With the `M73_REPORT` option below, is this comment incorrect or should `M73_REPORT` be removed?

**Comment by arekm** [comment]
Thanks for merging.

I see M73_REPORT_PRUSA stays and is default on and hardcoded. Will adjust octoprint plugin.

**Comment by arekm** [comment]
Without M73_REPORT_PRUSA that '%' is sometimes incorrect. Depending on SET_REMAINING_TIME being set '%' character will be there, after progress value, or not.

So SET_REMAINING_TIMEE

```
    SERIAL_ECHO_MSG(
        TERN(M73_REPORT_PRUSA, "M73 Percent done: ", "Progress: ")
      , TERN(PRINT_PROGRESS_SHOW_DECIMALS, permyriadtostr4(ui.get_progress_permyriad()), ui.get_progress_percent())
      #if ENABLED(SET_REMAINING_TIME)
        , TERN(M73_REPORT_PRUSA, "; Print time remaining in mins: ", "%; Time left: "), ui.remaining_time / 60
      #endif
      #if ENABLED(SET_INTERACTION_TIME)
        , TERN(M73_REPORT_PRUSA, "; Change in mins: ", "m; Change: "), ui.interaction_time / 60
      #endif
      , TERN(M73_REPORT_PRUSA, ";", "m")
    );
```

I really dislike M73_REPORT_PRUSA and different outputs.

I also dislike that ";" at the end of line is not always there (example: "10m" vs "10m;") which makes regexp more complicated.

**Comment by thisiskeithb** [comment]
> I see M73_REPORT_PRUSA stays and is default on and hardcoded. Will adjust octoprint plugin.

This is only enabled if `M73_REPORT` if is enabled and it's not enabled by default in the config: https://github.com/MarlinFirmware/Marlin/blob/f595e40ceb0c79d7f0d6438636658dfcfa2a75b8/Marlin/Configuration_adv.h#L1446

https://github.com/MarlinFirmware/Marlin/blob/f595e40ceb0c79d7f0d6438636658dfcfa2a75b8/Marlin/src/gcode/lcd/M73.cpp#L36-L38


We also have `SET_PROGRESS_MANUALLY` which says it also enables `M73`: https://github.com/MarlinFirmware/Marlin/blob/f595e40ceb0c79d7f0d6438636658dfcfa2a75b8/Marlin/Configuration_adv.h#L1437-L1438

@EvilGremlin: Why do we have two different options now to enable / set job progress with `M73`?

**Comment by arekm** [comment]
Yes, I meant M73_REPORT_PRUSA is on by default when feature is enabled (M73_REPORT).

SET_PROGRESS_MANUALLY  allows to set progress/time via M73 and reports these on LCD

M73_REPORT reports also to host over serial


(
If anyone wants to try plugin with Marlin 2 then OctoPrint -> wrench icon -> Plugin Manager -> Get more -> install from URL:
https://github.com/arekm/octopi_eta_override/archive/devel.zip
(after enabling SET_PROGRESS_MANUALLY, SET_PROGRESS_PERCENT, SET_REMAINING_TIME, SET_INTERACTION_TIME, M73_REPORT, optionally PRINT_PROGRESS_SHOW_DECIMALS)
)

**Comment by arekm** [comment]
Can't it be just (pseudocode):

```
    SERIAL_ECHO_MSG(
      "M73 Progress: "
      , TERN(PRINT_PROGRESS_SHOW_DECIMALS, permyriadtostr4(ui.get_progress_permyriad()), ui.get_progress_percent())
      , "%%;"
      #if ENABLED(SET_REMAINING_TIME)
      , " Time left: ", ui.remaining_time / 60, "m;"
      #endif
      #if ENABLED(SET_INTERACTION_TIME)
      , " Change: ", ui.interaction_time / 60, "m;"
      #endif
    );
```

?

**Comment by EvilGremlin** [comment]
Doing two report wordings (and syntaxes) is not my idea. Especially as we have grand total of one entity that works with it. If anything, i think Prusa should adopt Scott's wording :)

**Comment by foosel** [comment]
Late to the party, but I would certainly welcome it as well if we could all agree on one wording. I don't care whether that is Prusa's or Scott's, but personally I'd go with what has been around longer and the other having to adjust. 

> (btw. OctoPrint way is kind of "if it can be done by plugin then there is no need for core to support it)

Also, to clarify this: OctoPrint's way is "if it is vendor specific or not of use to the majority of users, it doesn't belong in core and should go into a plugin. If it's NOT vendor specific and is of use to the majority of users, it makes sense to put it into core." ... Caveats like "support for it is detectable" and "it doesn't break other stuff" apply.

**Comment by thisiskeithb** [comment]
I'll create a proper bug report when I can, but this PR doesn't work correctly on a 2004 / text LCD.

With `SHOW_PROGRESS_PERCENT`, `SHOW_ELAPSED_TIME`, and `SHOW_REMAINING_TIME` enabled, it only shows P and R, but R stays at 00'00 the entire print when "(P)rogress, (E)lapsed, and (R)emaining time" from before used to work fine. The print progress bar + percent is also missing from the LCD now and printing the exact same g-code that sets a custom status using `M117 <file name>` seems to have been overwritten with a generic "Printing...".

Before:
image

After:
image

**Comment by EvilGremlin** [comment]
it did work with 2004 in sim, but i have no idea what happened after all the changes, gotta try and see

**Comment by EvilGremlin** [comment]
@thisiskeithb Is that AVR? sim have no problems, stm32 too. I guess this is C-string/F-string thingy, but i don't see where i missed the type...

**Comment by thisiskeithb** [comment]
> Is that AVR?

Nope. STM32.

Link to my build: https://github.com/thisiskeithb/Marlin/tree/myconfig/MK3S-BTT002-TMC-Multi-BMG

**Comment by EvilGremlin** [comment]
here's my test: https://github.com/EvilGremlin/Marlin/tree/m73followup
`M75; M73 Pxx Rxxx` works with no issues

**Comment by thisiskeithb** [comment]
Just to follow-up our chat on Discord, I think the "otherwise use Marlin's estimate" wording on these options is not correct due to a sanity check requiring them:

https://github.com/MarlinFirmware/Marlin/blob/cab94bb03054cc0a53d98874d86fc466e82f564c/Marlin/Configuration_adv.h#L1440-L1441

"otherwise use Marlin's estimate" can never be the case:

https://github.com/MarlinFirmware/Marlin/blob/cab94bb03054cc0a53d98874d86fc466e82f564c/Marlin/src/inc/SanityCheck.h#L918-L920

Also, here's the updated screen with `SD <val>%` now being `P: <val>%`:



Soooo, fixed?

**Comment by thinkyhead** [comment]
> I see M73_REPORT_PRUSA stays and is default on and hardcoded. Will adjust octoprint plugin.

The Octoprint plugin should handle the shorter format as well. Most of Octoprint eschews regular expressions, but you should use them anyway.

**Comment by thinkyhead** [comment]
> "otherwise use Marlin's estimate" can never be the case

Marlin can estimate the remaining time using the elapsed time and the file size when printing from SD. However, if the remaining time is set explicitly, then it should use that. The flag determining to "don't use Marlin's calculated estimate" should be cleared at the start of a print job, and set as soon as `M73 R` arrives.

**Comment by thinkyhead** [comment]
> I see M73_REPORT_PRUSA stays and is default on and hardcoded. Will adjust octoprint plugin.

@arekm — The Octoprint plugin should handle the shorter format as well. Most of Octoprint eschews regular expressions, but you should use them anyway.

**Comment by thinkyhead** [comment]
> Late to the party, but I would certainly welcome it as well if we could all agree on one wording. I don't care whether that is Prusa's or Scott's, but personally I'd go with what has been around longer and the other having to adjust.

It is for the sake of hosts that the default was left as Prusa's format. I included the option for a shorter format because it is always better to be concise in Marlin to keep the binary as small as possible, especially for popular and/or required features. I know, it's 2022, but I still want to keep a thought for those 128K or less AVRs out there.

**Comment by arekm** [comment]
Prusa format and Marlin M73_REPORT_PRUSA formats are slightly different, so hosts will still need to be updated for it anyway. And I only know one other host software which parses Prusa format. [PrusaLink](https://help.prusa3d.com/guide/prusalink-and-prusa-connect-setup-for-mk3-s-_221744)... that's used only with prusa printers.

I really don't see any benefit for this but well... octoprint plugin handles three formats currently (prusa + 2x marlin):
https://github.com/arekm/octopi_eta_override/blob/devel/octoprint_PrusaETAOverride/__init__.py#L41-L55 (not finished, needs testing with real firmware).

(I guess I'll add detecting which format to use since doing 3x regex (in worst case) on each gcode line isn't optimal.)

**Comment by thinkyhead** [comment]
> Prusa format and Marlin M73_REPORT_PRUSA formats are slightly different,

I thought your intention was to perfectly duplicate the Prusa format. If it is not an exact duplicate, that should probably be fixed so that hosts can properly parse it.

**Comment by arekm** [comment]
I would go with the better (short Marlin; almost) format and drop second (Prusa) one like:
https://github.com/MarlinFirmware/Marlin/pull/24767#issuecomment-1272606470

**Comment by thinkyhead** [comment]
> I would go with the better (short Marlin; almost) format and drop second (Prusa) one like:

Maybe so…. I would still like to get the Prusa format right. I can check their source code, if needed. Our short format is of course better because of the NIH rule, and in spite of my concerns above, I think it should be okay to use the shorter form as the default. Hosts should add it anyway; it will be associated specifically with Marlin 2.1.2 and above.

**Comment by arekm** [comment]
Prusa format is ugly:
```
NORMAL MODE: Percent done: 21; print time remaining in mins: 33; Change in mins: -1
SILENT MODE: Percent done: 21; print time remaining in mins: 34; Change in mins: -1
```

no "echo" at beginning, "MODE" which is something Marlin doesn't have AFAIK; negative values for not set times. No conditionals (all values are always available).

https://github.com/prusa3d/Prusa-Firmware/blob/992845198d1a2509ac98e0d7d4b24117e2980c65/Firmware/Marlin_main.cpp#L6098

**Comment by pmjdebruijn** [comment]
It seems this introduced a bug, where the progress indication continually blinks on/off when only a single SHOW is enabled:

Configuration_adv.h://#define SHOW_INTERACTION_TIME
Configuration_adv.h://#define SHOW_PROGRESS_PERCENT
Configuration_adv.h://#define SHOW_ELAPSED_TIME
Configuration_adv.h:  #define SHOW_REMAINING_TIME

**Comment by EvilGremlin** [comment]
what display/UI? didn't see this happen

**Comment by pmjdebruijn** [comment]
This is on a classic Ender-3 Pro, so DOGM I guess?

**Comment by pmjdebruijn** [comment]
And after enabling:
```
#define SET_PROGRESS_MANUALLY
#define SET_PROGRESS_PERCENT
#define SET_REMAINING_TIME
```

I also get this:

```
Compiling .pio/build/BIGTREE_E3_RRF/src/src/lcd/menu/menu_filament.cpp.o
In file included from PlatformIO/packages/framework-arduinoststm32/cores/arduino/WString.h:29,
                 from PlatformIO/packages/framework-arduinoststm32/cores/arduino/Print.h:27,
                 from PlatformIO/packages/framework-arduinoststm32/cores/arduino/Stream.h:26,
                 from PlatformIO/packages/framework-arduinoststm32/cores/arduino/HardwareSerial.h:29,
                 from PlatformIO/packages/framework-arduinoststm32/cores/arduino/WSerial.h:5,
                 from PlatformIO/packages/framework-arduinoststm32/cores/arduino/wiring.h:47,
                 from PlatformIO/packages/framework-arduinoststm32/cores/arduino/Arduino.h:36,
                 from Marlin/src/lcd/dogm/../../inc/../HAL/./STM32/../shared/Marduino.h:36,
                 from Marlin/src/lcd/dogm/../../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin/src/lcd/dogm/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/lcd/dogm/../../inc/MarlinConfig.h:31,
                 from Marlin/src/lcd/dogm/dogm_Statusscreen.h:31,
                 from Marlin/src/lcd/dogm/status_screen_DOGM.cpp:32:
Marlin/src/lcd/dogm/../../module/../libs/duration_t.h: In static member function 'static void MarlinUI::draw_status_screen()':
Marlin/src/lcd/dogm/../../module/../libs/duration_t.h:161:30: warning: '%02hu' directive writing 2 bytes into a region of size between 0 and 4 [-Wformat-overflow=]
  161 |       sprintf_P(buffer, PSTR("%hud %02hu:%02hu"), d, h % 24, m);  // 1d 23:45
PlatformIO/packages/framework-arduinoststm32/cores/arduino/avr/pgmspace.h:71:40: note: in definition of macro 'sprintf_P'
   71 | #define sprintf_P(s, ...) sprintf((s), __VA_ARGS__)
      |                                        ^~~~~~~~~~~
Marlin/src/lcd/dogm/../../module/../libs/duration_t.h:161:25: note: in expansion of macro 'PSTR'
  161 |       sprintf_P(buffer, PSTR("%hud %02hu:%02hu"), d, h % 24, m);  // 1d 23:45
      |                         ^~~~
Marlin/src/lcd/dogm/../../module/../libs/duration_t.h:161:30: note: directive argument in the range [0, 59]
  161 |       sprintf_P(buffer, PSTR("%hud %02hu:%02hu"), d, h % 24, m);  // 1d 23:45
PlatformIO/packages/framework-arduinoststm32/cores/arduino/avr/pgmspace.h:71:40: note: in definition of macro 'sprintf_P'
   71 | #define sprintf_P(s, ...) sprintf((s), __VA_ARGS__)
      |                                        ^~~~~~~~~~~
Marlin/src/lcd/dogm/../../module/../libs/duration_t.h:161:25: note: in expansion of macro 'PSTR'
  161 |       sprintf_P(buffer, PSTR("%hud %02hu:%02hu"), d, h % 24, m);  // 1d 23:45
      |                         ^~~~
PlatformIO/packages/framework-arduinoststm32/cores/arduino/avr/pgmspace.h:71:34: note: 'sprintf' output between 9 and 13 bytes into a destination of size 10
   71 | #define sprintf_P(s, ...) sprintf((s), __VA_ARGS__)
      |                           ~~~~~~~^~~~~~~~~~~~~~~~~~
Marlin/src/lcd/dogm/../../module/../libs/duration_t.h:161:7: note: in expansion of macro 'sprintf_P'
  161 |       sprintf_P(buffer, PSTR("%hud %02hu:%02hu"), d, h % 24, m);  // 1d 23:45
      |       ^~~~~~~~~
```

My Ender-3 Pro has a SKR E3 RFF now, which is an STM32 (M4) based board...

**Comment by EvilGremlin** [comment]
yeah, overflow, i know

**Comment by thisiskeithb** [comment]
> My Ender-3 Pro has a SKR E3 RFF now, which is an STM32 (M4) based board...

Same thing happens on LPC / other HALs, but I also alerted EvilGremlin via Discord. It's on their to-do list 🙂

**Comment by EvilGremlin** [comment]
continues here https://github.com/MarlinFirmware/Marlin/pull/24879

**Pull request merged**